source/base/tensor_product_polynomials.cc

```diff
@@ -275,26 +275,77 @@ compute (const Point<dim>            &p,
+  // temporary memory allocation below does not pay off. Also we loop over all
+  // tensor polynomials in a peculiar way to avoid all dynamic memory allocation.
+  // We need to evaluate the polynomial value n_polynomials*dim times, and
+  // need to compute n_polynomials^dim tensor polynomials. Therefore we can split
+  // the loop over the tensor polynomials into one loop over n_polynomials,
+  // evaluate this polynomial for each dimension and multiply it with the
+  // associated n_polynomials^(dim-1) tensor polynomials before moving to the
+  // next polynomial.
+  if (n_values_and_derivatives == 1)
+    {
+      const unsigned int n_polynomials = polynomials.size();
+      const unsigned int factor = n_tensor_pols/n_polynomials;
+
+      std::fill(values.begin(),values.end(),1.0);
+
+      for (unsigned int polynomial=0; polynomial<n_polynomials; ++polynomial)
+        {
+          for (unsigned int d=0; d<dim; ++d)
+            {
+              const double value = polynomials[polynomial].value(p(d));
+
+              for (unsigned int f=0; f<factor; ++f)
+                {
+                  unsigned int index = 0;
+                  switch (d)
+                    {
+                    case 0:
+                      index = polynomial+f*n_polynomials;
+                      break;
+                    case 1:
+                      index = polynomial*n_polynomials + (f/n_polynomials)*n_polynomials*n_polynomials + f%n_polynomials;
+                      break;
+                    case 2:
+                      index = polynomial*factor + f;
+                      break;
+                    default:
+                      ExcNotImplemented();
+                      break;
+                    }
 
-  // compute the values (and derivatives, if
+                  const unsigned int i = index_map_inverse[index];
+                  values[i] *= value;
+                }
+            }
+        }
+      return;
+    }
+
+
+  // Compute the values (and derivatives, if
   // necessary) of all polynomials at this
-  // evaluation point. to avoid many
-  // reallocation, use one std::vector for
-  // polynomial evaluation and store the
-  // result as Tensor<1,5> (that has enough
+  // evaluation point. To avoid expensive memory allocation,
+  // we use a small amount of memory on the stack, and store the
+  // result in an array (that has enough
   // fields for any evaluation of values and
-  // derivatives, up to the 4th derivative)
-  Table<2,Tensor<1,5> > v(dim, polynomials.size());
-  {
-    std::vector<double> tmp (n_values_and_derivatives);
+  // derivatives, up to the 4th derivative, for up to 20 polynomials).
+  // If someone uses a larger number of
+  // polynomials, we need to allocate more memory on the heap.
+  double (*v)[dim][5];
+  double small_array[20][dim][5];
+
+  const unsigned int n_polynomials = polynomials.size();
+  if (n_polynomials > 20)
+    v = new double[n_polynomials][dim][5];
```

Comment:
I feel a bit uncomfortable with `new/delete` constructions when there are function calls inside such as `polynomials[i].value(...)` that could go wrong. In that case, the array will not be cleaned up and you have a memory leak. However, you cannot put `double[dim][5]` into a `std::vector` straight away, but rather need something like `std::array<std::array<double,5>,dim>` which might complicate things and probably be more expensive as your experiments with the old code involving a table of tensors indicated. My suggestion would be to create `small_array[20*dim*5]` and a `std::vector<double> large_array` that is resized (if `n_polynomials>20`) to `large_array.resize(n_polynomials*dim*5)`. You will need to do some indexing `i*dim*5+d*5+...` that the compiler does for you otherwise but I'd still prefer to avoid explicit `new/delete` statements.